packages/flutter/lib/src/foundation/tree_diagnostics_mixin.dart

```diff
@@ -2,9 +2,1438 @@
+  /// `showNull` indicates whether the property should still be shown if its
+  /// `value` is null. `hidden` specifies whether the property should be hidden
+  /// or shown. `nullDescription` provides a description of the property value
+  /// if it is null.
+  /// If specified, exactly `fractionDigits` after the decimal point are shown.
+  /// If specified, `unit` describes the unit for the value (e.g. px).
+  factory DiagnosticsNode.doubleProperty(String name, double value, {
+    bool showNull: true,
+    int fractionDigits,
+    bool hidden: false,
+    String nullDescription,
+    String unit,
+  }) {
+    String description;
+    if (value == null) {
+      description = nullDescription ?? value.toString();
+    } else {
+      description = fractionDigits != null ?
+         value.toStringAsFixed(fractionDigits) : value.toString();
+    }
+
+    if (unit != null)
+      description = '$description$unit';
+
+    return new _LeafDiagnosticsValue(
+      name: name,
+      header: description,
+      object: value,
+      showNull: showNull,
+      hidden: hidden,
+    );
+  }
+
+  /// Property with a [double] value where computing its value might throw an
+  /// exception.
+  ///
+  /// If an exception is thrown, the value is set to null and the description
+  /// indicates what exception was thrown.
+  /// If specified, exactly `fractionDigits` after the decimal point are shown.
+  factory DiagnosticsNode.unsafeDoubleProperty(
+      String name, double computeValue(), { int fractionDigits }) {
+    try {
+      return new DiagnosticsNode.doubleProperty(name, computeValue(), fractionDigits: fractionDigits);
+    } catch (e) {
+      // TODO(jacobr): track the exception thrown as part of the Diagnostics node.
+      return new DiagnosticsNode.doubleProperty(
+        name,
+        null,
+        fractionDigits: fractionDigits,
+        nullDescription: 'EXCEPTION (${e.runtimeType})',
+      );
+    }
+  }
+
+
+  /// A property describing a double value as a percent.
+  /// The percentage is clamped between 0 and 100% inclusive.
+  factory DiagnosticsNode.percentProperty(String name, double fraction, {
+    String nullDescription,
+    bool showName: true,
+  }) {
+    return new _LeafDiagnosticsValue(
+      name: name,
+      header: fraction != null ?
+          '${(fraction.clamp(0.0, 1.0) * 100.0).toStringAsFixed(1)}%' :
+          nullDescription,
+      object: fraction,
+      showName: showName,
+    );
+  }
+
+
+  /// A property describing whether a callback function is present.
+  factory DiagnosticsNode.callbackProperty(String name, Function callback, {
+    @required String ifPresent,
+  }) {
+    return new _LeafDiagnosticsValue(
+      name: name,
+      object: callback,
+      header: callback != null ? ifPresent : 'null',
+      hidden: callback == null,
+      showName: false,
+    );
+  }
+
+  /// A property describing that a callback function is not present.
```

Comment:
This could be combined with callbackProperty into a single constructor that takes an "ifPresent" and an "ifMissing" (and requires one or both to be present).